### PR TITLE
Router

### LINKED ISSUE #262: Way to switch intents?
Is there a way to trigger/switch an intent manually? 

My use case:
An intent is triggered. 
Logic is performed that decides if a user has access to this particular intent. 
If not, then pass the user to another intent on the alexa service and continue the interaction.

### PR DESCRIPTION
**Pull request by fremail**
Hello!

I implemented a router requested in #262.
Actually I brought most code of the router from the incoming request routing based on request type.

In two words, I did two things:

1) Added ability to route request to intent/AudioPlayer event handler/PlaybackController event handler/custom request handler and other request types.
`request.getRouter().launch();`
`request.getRouter().sessionEnded();`
`request.getRouter().intent("oneIntent");`
`request.getRouter().custom("sayHello");`
`request.getRouter().audioPlayer("PlaybackNearlyFinished");`
`request.getRouter().playbackController("NextCommandIssued");`
`request.getRouter().displayElementSelected();`

2) Simplify incoming request routing by using the new router.

resolves #262, #102

### CODE REVIEW

**Comment by fremail** [comment]
@kobim AFAIK Request will refer to Router (not have a copy of it), so there shouldn't be memory issues.

I did similar to Session object which is in Response. But the Router is in Request because it sounds better: "Route **request** to ...".

There is no problem to separate it from Request if you want.
_________

Also during last several days I'm thinking about some flag for stopping next routing.
For example in `pre` you decide to switch to `myIntent` intent though the original request was to `originalIntent`. And you do not want to execute `originalIntent`. My example in code:
```
// incoming request is to originalIntent
app.pre = function (request, response) {
    return request.getRouter().intent('myIntent').end();
};
app.intent('originalIntent', function () {
    // won't be executed in this example
});
app.intent('myIntent', function (request, response) {
    response.say('Hi there!');
});
```
Though maybe it's worth to be a separate PR.

**Comment by kobim** [comment]
@dblock what do you think?

**Comment by dblock** [comment]
This looks good to me. I say @lazerwalker makes the final call :)

**Comment by lazerwalker** [comment]
Sorry, somehow missed the earlier ping for this! This all LGTM 👍. 

I'm not too concerned about memory leaks — I've never seen circular references like that cause any practical JS issues, and some Googling suggests modern JS engine GCs are generally smart enough to handle situations like that without trouble.

It felt weird at first glance, but thinking about it, I think having router be an attribute on request makes more sense than alternatives. I like how it mirrors how we handle session.

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@
 * [#364](https://github.com/alexa-js/alexa-app/pull/364): Fix reprompt() to concatenate multiple SSML prompts - [@andrewjhunt](https://github.com/andrewjhunt).
 * [#371](https://github.com/alexa-js/alexa-app/pull/371): Call response.prepare() after post() - [@fremail](https://github.com/fremail).
 * [#373](https://github.com/alexa-js/alexa-app/pull/373): Keep multiple newlines in card output - [@kielni](https://github.com/kielni) and [@fremail](https://github.com/fremail).
+* [#374](https://github.com/alexa-js/alexa-app/pull/374): Router - [@fremail](https://github.com/fremail).
```
Comment (dblock) [inline CHANGELOG.md]: Can we change this to "Added Router" or something more detailed?

**Comment by dblock** [comment]
I made a nitpick on changelog, otherwise anyone hit merge pls.

**Comment by fremail** [comment]
Updated Changelog.md.

Squashed and rebased my commits.

**Pull request merged**